experimental/devices/ads1x15/ads1x15.go

```diff
@@ -0,0 +1,392 @@
+	config |= gainConf
+	// Set the mode (continuous or single shot).
+	config |= ads1x15ConfigModeSingle
+
+	// Set the data rate (this is controlled by the subclass as it differs
+	// between ADS1015 and ADS1115).
+	config |= dataRateConf
+	config |= ads1x15ConfigCompQueDisable // Disable comparator mode.
+
+	// Build the query to the ADC
+	configBytes := make([]byte, 2)
+	binary.BigEndian.PutUint16(configBytes, config)
+	preparedQuery := append([]byte{ads1x15PointerConfig}, configBytes...)
+
+	// The wait for the ADC sample to finish is based on the sample rate plus a
+	// small offset to be sure (0.1 millisecond).
+	waitTime := time.Duration(1000000/dataRate+100) * time.Microsecond
+
+	pin = &ads1x15AnalogPin{
+		adc:               d,
+		preparedQuery:     preparedQuery,
+		voltageMultiplier: voltageMultiplier,
+		waitTime:          waitTime,
+	}
+
+	return
+}
+
+func (d *Dev) executePreparedQuery(query []byte, waitTime time.Duration, voltageMultiplier physic.ElectricPotential) (value physic.ElectricPotential, err error) {
+	// Lock the ADC converter to avoid multiple simultaneous readings.
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// Send the config value to start the ADC conversion.
+	// Explicitly break the 16-bit value down to a big endian pair of bytes.
+	if err = d.c.Tx(query, nil); err != nil {
+		return
+	}
+
+	// Wait for the ADC sample to finish.
+	time.Sleep(waitTime)
+
+	// Retrieve the result.
+	data := []byte{0, 0}
+	if err = d.c.Tx([]byte{ads1x15PointerConversion}, data); err != nil {
+		return
+	}
+
+	// Convert the raw data into physical value.
+	result := physic.ElectricPotential(binary.BigEndian.Uint16(data))
+
+	if result&0x8000 != 0 {
+		result -= 1 << 16
+	}
+
+	value = result * voltageMultiplier / physic.ElectricPotential(1<<15)
+
+	return
+}
+
+// bestGainForElectricPotential returns the gain the most adapted to read up to the specified difference of potential.
+func (d *Dev) bestGainForElectricPotential(voltage physic.ElectricPotential) (bestGain int, err error) {
+	var max physic.ElectricPotential
+	difference := physic.ElectricPotential(math.MaxInt64)
+	currentBestGain := -1
+
+	for key, value := range d.gainVoltage {
+		// We compute the maximum in case we need to display an error
+		if value > max {
+			max = value
+		}
+		newDiff := value - voltage
+		if newDiff >= 0 && newDiff < difference {
+			difference = newDiff
+			currentBestGain = key
+		}
+	}
+
+	if currentBestGain < 0 {
+		err = fmt.Errorf("The maximum voltage which can be read is %s", max.String())
+		return
+	}
+
+	bestGain = currentBestGain
+	return
+}
+
+// bestDataRateForFrequency returns the gain the most data rate to read samples at least at the requested frequency.
+func (d *Dev) bestDataRateForFrequency(minimumFrequency physic.Frequency) (bestDataRate int, err error) {
+	var max physic.Frequency
+	difference := physic.Frequency(math.MaxInt64)
+	currentBestDataRate := -1
+
+	for key := range d.dataRates {
```

Comment:
Issues:
- You are not going through the map in order, this is problematic
- So I recommend to convert the map into an ordered slice of struct.

```
type dataRate struct {
  f physic.Frequency
  v uint16
}
 
var ads1015DataRates = [...]dataRate{
  {128*physic.Hertz, 0},
  {250*physic.Hertz, 0x20},
  ...
}
```

Then you can do:
```
for i := d.dataRates {
  if f <= d.dataRates[i].f {
    return d.dataRates[i].v
  }
}
return d.dataRates[len(d.dataRates)-1].v
```
This is much simpler. You could return an error if the datarate is too high.